File: client/python/src/transport_server_client/video/factory.py
"""
Factory functions for creating RobotHub TransportServer video clients
"""

from .consumer import VideoConsumer
from .producer import VideoProducer
from .types import ClientOptions, ParticipantRole


def create_client(
    role: ParticipantRole,
    base_url: str,
    options: ClientOptions | None = None,
) -> "VideoProducer | VideoConsumer":
    """Factory function to create the appropriate client based on role"""
    if role == ParticipantRole.PRODUCER:
        return VideoProducer(base_url, options)
    if role == ParticipantRole.CONSUMER:
        return VideoConsumer(base_url, options)
    msg = f"Invalid role: {role}. Must be 'producer' or 'consumer'"
    raise ValueError(msg)


async def create_producer_client(
    base_url: str,
    workspace_id: str | None = None,
    room_id: str | None = None,
    participant_id: str | None = None,
    options: ClientOptions | None = None,
) -> VideoProducer:
    """Create and connect a producer client"""
    producer = VideoProducer(base_url, options)

    workspace_id, room_id = await producer.create_room(workspace_id, room_id)
    connected = await producer.connect(workspace_id, room_id, participant_id)

    if not connected:
        msg = "Failed to connect as video producer"
        raise ValueError(msg)

    return producer


async def create_consumer_client(
    workspace_id: str,
    room_id: str,
    base_url: str,
    participant_id: str | None = None,
    options: ClientOptions | None = None,
) -> VideoConsumer:
    """Create and connect a consumer client"""
    consumer = VideoConsumer(base_url, options)
    connected = await consumer.connect(workspace_id, room_id, participant_id)

    if not connected:
        msg = "Failed to connect as video consumer"
        raise ValueError(msg)

    return consumer

File: client/js/src/video/factory.ts
/**
 * Factory functions for creating RobotHub TransportServer video clients
 */

import { VideoProducer } from './producer.js';
import { VideoConsumer } from './consumer.js';
import type { ParticipantRole, ClientOptions } from './types.js';

/**
 * Factory function to create the appropriate client based on role
 */
export function createClient(
  role: ParticipantRole,
  baseUrl: string,
  options: ClientOptions = {}
): VideoProducer | VideoConsumer {
  if (role === 'producer') {
    return new VideoProducer(baseUrl, options);
  }
  if (role === 'consumer') {
    return new VideoConsumer(baseUrl, options);
  }
  throw new Error(`Invalid role: ${role}. Must be 'producer' or 'consumer'`);
}

/**
 * Create and connect a producer client
 */
export async function createProducerClient(
  baseUrl: string,
  workspaceId?: string,
  roomId?: string,
  participantId?: string,
  options: ClientOptions = {}
): Promise<VideoProducer> {
  const producer = new VideoProducer(baseUrl, options);

  const roomData = await producer.createRoom(workspaceId, roomId);
  const connected = await producer.connect(roomData.workspaceId, roomData.roomId, participantId);

  if (!connected) {
    throw new Error('Failed to connect as video producer');
  }

  return producer;
}

/**
 * Create and connect a consumer client
 */
export async function createConsumerClient(
  workspaceId: string,
  roomId: string,
  baseUrl: string,
  participantId?: string,
  options: ClientOptions = {}
): Promise<VideoConsumer> {
  const consumer = new VideoConsumer(baseUrl, options);
  const connected = await consumer.connect(workspaceId, roomId, participantId);

  if (!connected) {
    throw new Error('Failed to connect as video consumer');
  }

  return consumer;
} 